src/ucp/api/ucp.h

```diff
@@ -582,16 +582,19 @@ enum ucp_am_cb_flags {
  * routines.
  */
 enum ucp_send_am_flags {
-    UCP_AM_SEND_FLAG_REPLY = UCS_BIT(0),             /**< Force relevant reply
-                                                          endpoint to be passed to
-                                                          the data callback on the
-                                                          receiver. */
-    UCP_AM_SEND_FLAG_EAGER = UCS_BIT(1),             /**< Force UCP to use only
-                                                          eager protocol for AM sends. */
-    UCP_AM_SEND_FLAG_RNDV  = UCS_BIT(2),             /**< Force UCP to use only
-                                                          rendezvous protocol for
-                                                          AM sends. */
-    UCP_AM_SEND_REPLY      = UCP_AM_SEND_FLAG_REPLY  /**< Backward compatibility. */
+    UCP_AM_SEND_FLAG_REPLY       = UCS_BIT(0),             /**< Force relevant reply
+                                                                endpoint to be passed to
+                                                                the data callback on the
+                                                                receiver. */
+    UCP_AM_SEND_FLAG_EAGER       = UCS_BIT(1),             /**< Force UCP to use only
+                                                                eager protocol for AM sends. */
+    UCP_AM_SEND_FLAG_RNDV        = UCS_BIT(2),             /**< Force UCP to use only
+                                                                rendezvous protocol for
+                                                                AM sends. */
+    UCP_AM_SEND_FLAG_TEMP_HEADER = UCS_BIT(3),             /**< Signal UCX that this header
```

Comment:
IMO need to find a better name for it..

Reply:
Story of my life...
Do you have any idea?

Reply:
UCP_AM_SEND_FLAG_COPY_HEADER
The user header buffer is copied-out during the active message send routine, and can be released or invalidated when the send routing returns, even if the send request is not completed yet.
@shamisp WDYT?

Reply:
IMO shouldn't put "COPY" there, Because UCX doesn't **always** copy the header (but only when needed).
Maybe the flag name should reflect that the user can release it / UCX doesn't need it, rather than misleading to think that it is "always copied".

Reply:
actually, it **is** always copied - either by UCP or UCT layer. Even when put immediately on the QP, we copy it to UCT bounce buffer.
It it wasn't copied to another location, we had to wait at least until the data is sent to the remote peer - which we don't.

Reply:
@ofirfarjun7 Do you have reference implementation that you can share ? It will help me to understand what you are trying to say here.

Reply:
Hi.
I submitted a PR of the implementation, currently it's marked as draft but you can check it (#8173).
I've also added a flowchart describing the send operation in UCP layer with the changes I added, look again (just added it) in the PR description for the link, I hope you have access to this directory.

Reply:
👍
So I remove the link...

Reply:
Hi.
As Leonid mentioned I don't think you have access to our cloud so I removed the link.

In short - ucp_am_send_nbx can return success or in progress.
In case it's not succeed immediately it goes to the pending queue.
In this case the provided user header is still not part of the future srq, and therefore if the user release/corrupt it before progressing the workers, and retry sending the message bad things can happen...

ucp_send_nbx -> success -> user can reuse/release buffer
       |
      \/
in progress -> send request not posted to HW, user must keep header persistent until progressing the worker and complete the send operation.

We want to change it, in case the send operation didn't complete we will store the header in UCX internal buffer.
This is relevant for short/bcopy/RNDV.
In zcopy we already copy the buffer to registered UCP buffer.

Reply:
@ofirfarjun7 great job on diagram.

Reply:
UCP_AM_SEND_FLAG_COPY_HEADER ->UCP_AM_SEND_FLAG_CACHE_HEADER

The flag indicates that the header must be cached by UCP library. Once the ucp_send_am_* functions returns the header buffer can be reused or released.

Reply:
I think this name is better than all we come with until now

Reply:
BTW, I think this should be default behavior to start with. It is too late to change the behavior now so the flag is really welcome. The implementation looks surprisingly complex...

Reply:
Yes. it's surprise me too...
The difference between short and the bcopy complicated the implementation...
In short the pack CB is called before we know if we have enough resources because we use iov API.
In bcopy the pack CB is called after we know we have enough resources.

If you have any idea how to simplify it I will be happy to hear it.